File: src/test/rustdoc-js/generics.js
// exact-check

const QUERY = [
    'R<P>',
    '"P"',
    'P',
    'ExtraCreditStructMulti<ExtraCreditInnerMulti, ExtraCreditInnerMulti>',
    'TraitCat',
    'TraitDog',
    'Result<String>',
];

const EXPECTED = [
    {
        // R<P>
        'returned': [
            { 'path': 'generics', 'name': 'alef' },
        ],
        'in_args': [
            { 'path': 'generics', 'name': 'alpha' },
        ],
    },
    {
        // "P"
        'others': [
            { 'path': 'generics', 'name': 'P' },
        ],
        'returned': [
            { 'path': 'generics', 'name': 'alef' },
        ],
        'in_args': [
            { 'path': 'generics', 'name': 'alpha' },
        ],
    },
    {
        // P
        'returned': [
            { 'path': 'generics', 'name': 'alef' },
        ],
        'in_args': [
            { 'path': 'generics', 'name': 'alpha' },
        ],
    },
    {
        // "ExtraCreditStructMulti"<ExtraCreditInnerMulti, ExtraCreditInnerMulti>
        'in_args': [
            { 'path': 'generics', 'name': 'extracreditlabhomework' },
        ],
        'returned': [],
    },
    {
        // TraitCat
        'in_args': [
            { 'path': 'generics', 'name': 'gamma' },
        ],
    },
    {
        // TraitDog
        'in_args': [
            { 'path': 'generics', 'name': 'gamma' },
        ],
    },
    {
        // Result<String>
        'others': [],
        'returned': [
            { 'path': 'generics', 'name': 'super_soup' },
        ],
        'in_args': [
            { 'path': 'generics', 'name': 'super_soup' },
        ],
    },
];

File: src/test/rustdoc-js/generics.rs
pub struct P;
pub struct Q;
pub struct R<T>(T);

// returns test
pub fn alef() -> R<P> { loop {} }
pub fn bet() -> R<Q> { loop {} }

// in_args test
pub fn alpha(_x: R<P>) { loop {} }
pub fn beta(_x: R<Q>) { loop {} }

// test case with multiple appearances of the same type
pub struct ExtraCreditStructMulti<T, U> { t: T, u: U }
pub struct ExtraCreditInnerMulti {}
pub fn extracreditlabhomework(
    _param: ExtraCreditStructMulti<ExtraCreditInnerMulti, ExtraCreditInnerMulti>
) { loop {} }
pub fn redherringmatchforextracredit(
    _param: ExtraCreditStructMulti<ExtraCreditInnerMulti, ()>
) { loop {} }

pub trait TraitCat {}
pub trait TraitDog {}

pub fn gamma<T: TraitCat + TraitDog>(t: T) {}
